### PR TITLE
Core Implementation

### PR DESCRIPTION
**Pull request by michavie**
Migrating from https://github.com/PeerMeHQ/mx-sdk-php

Dependencies:
- `bitwasp/bech32`:  [Repository](https://github.com/Bit-Wasp/bech32) | [Distribution](https://packagist.org/packages/bitwasp/bech32) | Installs: 465k
- `brick/math`:  [Repository](https://github.com/brick/math) | [Distribution](https://packagist.org/packages/brick/math) | Installs: 214.9m
- `kornrunner/keccak`:  [Repository](https://github.com/kornrunner/php-keccak) | [Distribution](https://packagist.org/packages/kornrunner/keccak) | Installs: 1.29m
- `illuminate/collections`:  [Repository](https://github.com/illuminate/collections) | [Distribution](https://packagist.org/packages/illuminate/collections) | Installs: 26.4m

### CODE REVIEW

File: composer.json
Comment (andreibancioiu) [inline composer.json:20]: For other SDKs, we tend to use fixed versions for the dependencies. Should we do here the same?

File: composer.json
Comment (andreibancioiu) [inline composer.json]: Pending comment, until checking cryptographic utilities:

How about using the one bundled with PHP?
https://www.php.net/manual/en/book.sodium.php

File: composer.json
Comment (andreibancioiu) [reply composer.json:20]: Also, do you think you can list the dependencies in the PR description, along with a GitHub URL, and possibly the number of stars and the date of the most recent release?

File: src/Address.php
Comment (andreibancioiu) [inline src/Address.php]: For future PR: constants can be extracted, where applicable / where it makes sense to do so.

File: src/Address.php
Comment (andreibancioiu) [inline src/Address.php]: Pending comment: double-check with colleagues if this is the standard pattern in PHP. Can also stay as it is for now.

File: tests/Unit/AddressTest.php
Comment (andreibancioiu) [inline tests/Unit/AddressTest.php:1]: If possible, use test vectors from `sdk-js` and `sdk-py` (here and in other tests).

File: src/Constants.php
Comment (andreibancioiu) [inline src/Constants.php:9]: OK for now. When aligning the SDK to the specs, this will change (e.g. a transactions factory config will be used).

File: tests/Unit/UserSignerTest.php
Comment (andreibancioiu) [inline tests/Unit/UserSignerTest.php:19]: Can we use a data vector from JS or PY - just to make sure this does not serve as a slightly misleading example? Generally speaking, users should not be signing plain strings, but properly serialized messages, instead - e.g. `SignableMessage`.

File: src/Utils/Encoder.php
Comment (andreibancioiu) [inline src/Utils/Encoder.php]: A bit tricky - aren't there false positives sometimes?

File: src/Utils/Encoder.php
Comment (andreibancioiu) [inline src/Utils/Encoder.php]: Perhaps users should never set `skipHexBytePadding = true`?

File: src/Utils/Decoder.php
Comment (andreibancioiu) [inline src/Utils/Decoder.php:21]: Perhaps add more test vectors in `DecoderTest.php`?

File: src/UserSecretKey.php
Comment (andreibancioiu) [inline src/UserSecretKey.php]: Maybe also include a release tag in the link (since this might change upon alignment to specs).

**Review by andreibancioiu — COMMENTED** [review]
Awesome :tada:

File: composer.json
Comment (michavie) [reply composer.json:20]: Using fixed versions is not standard practice for defining composer dependencies because it is highly probable for package consumers to face version conflicts given how composer resolves project dependencies:

_Composer (PHP) and NPM (Node.js) handle multiple versions of dependencies differently. Composer installs a single version of a package per project, focusing on dependency compatibility within that project. It resolves dependency conflicts to ensure that only one compatible version of each package is used._

---

I listed the dependencies with info in the PR description

File: src/Utils/Encoder.php
Comment (michavie) [reply src/Utils/Encoder.php]: Removed; stemmed from old times

File: src/Utils/Encoder.php
Comment (michavie) [reply src/Utils/Encoder.php]: Good catch, potentially improved

File: src/Address.php
Comment (michavie) [reply src/Address.php]: Good point; no more magic numbers

File: src/Address.php
Comment (michavie) [reply src/Address.php]: Changed to mimic implementation of sdk-js

File: src/UserSecretKey.php
Comment (michavie) [reply src/UserSecretKey.php]: Done

File: composer.json
Comment (michavie) [reply composer.json]: Good point! Updated the implementation – should be safe to use given that sodium utilities are bundled with PHP core starting v7.2, and the min PHP version for this package is above

Removed originally included alternative dependency

File: src/Utils/Decoder.php
Comment (michavie) [reply src/Utils/Decoder.php:21]: Added

File: tests/Unit/UserSignerTest.php
Comment (michavie) [reply tests/Unit/UserSignerTest.php:19]: For `UserVerifier` and `UserSigner` there is currently no test coverage for these implementations in neither `js` nor `py` SDKs. I started to minimally adapt method signatures to the new specs which accept `bytes`, and abstracted `bytes` through `Bytes` value object

File: tests/Unit/AddressTest.php
Comment (michavie) [reply tests/Unit/AddressTest.php:1]: Done to a wider extent while seeking not to replicate tests that will become obsolete with the coming updates for the new specs

File: src/UserSecretKey.php
Comment (andreibancioiu) [inline src/UserSecretKey.php:27]: :+1:

File: src/Address.php
Comment (andreibancioiu) [inline src/Address.php]: I think this should be called: `BECH32_ADDRESS_LENGTH`.

File: tests/Unit/TransactionTest.php
Comment (andreibancioiu) [inline tests/Unit/TransactionTest.php:100]: :rocket:

**Review by andreibancioiu — DISMISSED** [review]
Awesome :100: 

Perhaps we can also add in README a short note about how to run the unit tests etc.

File: src/Address.php
Comment (michavie) [reply src/Address.php]: True, renamed

**Comment by michavie** [comment]
@andreibancioiu 

Added minimal readme as part of this PR as well. Will be expanded as more parts are added.

Other pieces of the migration like network components, GitHub workflows & co will follow in separate PR

**Review by andreibancioiu — APPROVED** [review]

**Comment by razvanphp** [comment]
@andreibancioiu I see this required 2 approvals, can I be the second maintainer? I will test this directly in my API code. Thanks!

File: composer.json
Comment (razvanphp) [reply composer.json:20]: Patch version should always be safe to upgrade, no breaking changes expected.

**Comment by andreibancioiu** [comment]
> @andreibancioiu I see this required 2 approvals, can I be the second maintainer? I will test this directly in my API code. Thanks!

@razvanphp, added :tada: 

Sorry for the delay :pray:

**Comment by razvanphp** [comment]
Thanks! Should we have an oficial composer account linked to the repo?

**Review by razvanphp — APPROVED** [review]

**Pull request merged**
